File: apps/api/src/storage/storage.service.ts
// Local-filesystem storage service. Replaces Supabase Storage.
//
// Files are written under STORAGE_ROOT/<bucket>/<pathPrefix>/<uuid><ext>.
// The storage path returned to callers ("<bucket>/<prefix>/<uuid>.ext") is
// the same shape the rest of the app already stores in *_path columns,
// so no caller changes are needed apart from how files are read back.
//
// Reading is exposed via StorageService.read(path) — the verify and deed
// download endpoints stream the bytes through, never expose the FS path.

import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { mkdir, writeFile, readFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { dirname, extname, join, resolve } from 'node:path';
import { randomUUID } from 'node:crypto';
import { SijilliErrors } from '../common/errors/error-envelope';
import { sha256Hex } from '../common/utils/sha256';

export interface UploadFile {
  originalname: string;
  mimetype: string;
  size: number;
  buffer: Buffer;
}

export interface UploadOptions {
  bucket: string;
  pathPrefix: string;
  maxBytes: number;
  allowedMimeTypes: readonly string[];
}

export interface UploadResult {
  bucket: string;
  path: string;
  size: number;
  mimeType: string;
  sha256: string;
}

@Injectable()
export class StorageService implements OnModuleInit {
  private readonly logger = new Logger(StorageService.name);
  private root!: string;

  constructor(private readonly config: ConfigService) {}

  onModuleInit() {
    const dir = this.config.get<string>('STORAGE_ROOT') ?? join(process.cwd(), 'storage');
    this.root = resolve(dir);
    if (!existsSync(this.root)) {
      // We can't await in onModuleInit, but mkdir(recursive) on first
      // upload will create what's needed. Just log the chosen root.
      this.logger.log(`Storage root will be created on first upload: ${this.root}`);
    } else {
      this.logger.log(`Storage root: ${this.root}`);
    }
  }

  async upload(file: UploadFile, opts: UploadOptions): Promise<UploadResult> {
    if (!file || !file.buffer || file.size === 0) {
      throw SijilliErrors.validation('الملف فارغ أو غير صالح.', 'Empty or invalid file.');
    }
    if (file.size > opts.maxBytes) {
      throw SijilliErrors.validation(
        `حجم الملف (${formatBytes(file.size)}) يتجاوز الحد المسموح (${formatBytes(opts.maxBytes)}).`,
        `File size ${file.size} exceeds limit ${opts.maxBytes}.`,
      );
    }
    if (!opts.allowedMimeTypes.includes(file.mimetype)) {
      throw SijilliErrors.validation(
        `نوع الملف "${file.mimetype}" غير مدعوم.`,
        `Unsupported mime type ${file.mimetype}.`,
      );
    }

    const ext = sanitizeExtension(file.originalname, file.mimetype);
    const objectName = `${randomUUID()}${ext}`;
    const relPath = `${opts.pathPrefix.replace(/\/+$/, '')}/${objectName}`;
    const absPath = this.absoluteFor(opts.bucket, relPath);

    await mkdir(dirname(absPath), { recursive: true });
    await writeFile(absPath, file.buffer, { flag: 'wx' });

    return {
      bucket: opts.bucket,
      path: relPath,
      size: file.size,
      mimeType: file.mimetype,
      sha256: sha256Hex(file.buffer),
    };
  }

  // Read a previously-uploaded file. Throws notFound if missing.
  async read(bucket: string, path: string): Promise<Buffer> {
    const abs = this.absoluteFor(bucket, path);
    if (!existsSync(abs)) throw SijilliErrors.notFound('الملف');
    return readFile(abs);
  }

  // Write raw bytes (used by the deed generator for already-rendered PDFs).
  // Overwrites if the path exists.
  async writeRaw(
    bucket: string,
    path: string,
    data: Buffer | Uint8Array,
    mime: string,
  ): Promise<UploadResult> {
    const abs = this.absoluteFor(bucket, path);
    await mkdir(dirname(abs), { recursive: true });
    const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
    await writeFile(abs, buf);
    return {
      bucket,
      path,
      size: buf.byteLength,
      mimeType: mime,
      sha256: sha256Hex(buf),
    };
  }

  // Resolve, refusing path traversal. The bucket and path are joined under
  // root and re-resolved; if the result escapes root, reject.
  private absoluteFor(bucket: string, path: string): string {
    const safeBucket = bucket.replace(/[^a-zA-Z0-9_-]/g, '_');
    const candidate = resolve(this.root, safeBucket, path);
    if (!candidate.startsWith(resolve(this.root, safeBucket))) {
      throw SijilliErrors.validation('مسار غير صالح.', 'Invalid path.');
    }
    return candidate;
  }
}

function sanitizeExtension(originalName: string, mimeType: string): string {
  const ext = extname(originalName).toLowerCase();
  if (/^\.[a-z0-9]{1,6}$/.test(ext)) return ext;
  switch (mimeType) {
    case 'application/pdf':
      return '.pdf';
    case 'image/jpeg':
      return '.jpg';
    case 'image/png':
      return '.png';
    default:
      return '.bin';
  }
}

function formatBytes(n: number): string {
  if (n < 1024) return `${n} B`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} KB`;
  return `${(n / 1024 / 1024).toFixed(1)} MB`;
}

File: apps/api-dotnet/Storage/StorageService.cs
using System.Security.Cryptography;
using System.Text.RegularExpressions;
using Sijilli.Api.Common.Errors;

namespace Sijilli.Api.Storage;

// Local-filesystem storage. Replaces Supabase Storage. Files live under
// STORAGE_ROOT/<bucket>/<pathPrefix>/<uuid><ext>. The string returned to
// callers ("<bucket>/<prefix>/<uuid>.ext") matches the existing *_path
// columns, so no DB shape changes.
//
// Reading flows through StorageService.ReadAsync — verify endpoints stream
// the bytes through and never expose the FS path.
public sealed partial class StorageService
{
    private static readonly Regex SafeBucketRe = new("[^a-zA-Z0-9_-]");
    private static readonly Regex SafeExtRe = new(@"^\.[a-z0-9]{1,6}$");

    private readonly string _root;
    private readonly ILogger<StorageService> _log;

    public StorageService(IConfiguration config, ILogger<StorageService> log)
    {
        _log = log;
        var dir = config["Sijilli:StorageRoot"]
            ?? Environment.GetEnvironmentVariable("STORAGE_ROOT")
            ?? Path.Combine(Directory.GetCurrentDirectory(), "storage");
        _root = Path.GetFullPath(dir);
        if (!Directory.Exists(_root))
            log.LogInformation("Storage root will be created on first write: {Root}", _root);
        else
            log.LogInformation("Storage root: {Root}", _root);
    }

    public sealed class UploadFile
    {
        public required string OriginalName { get; init; }
        public required string MimeType { get; init; }
        public required long Size { get; init; }
        public required byte[] Buffer { get; init; }
    }

    public sealed class UploadOptions
    {
        public required string Bucket { get; init; }
        public required string PathPrefix { get; init; }
        public required long MaxBytes { get; init; }
        public required IReadOnlyList<string> AllowedMimeTypes { get; init; }
    }

    public sealed class UploadResult
    {
        public required string Bucket { get; init; }
        public required string Path { get; init; }
        public required long Size { get; init; }
        public required string MimeType { get; init; }
        public required string Sha256 { get; init; }
    }

    public async Task<UploadResult> UploadAsync(UploadFile file, UploadOptions opts, CancellationToken ct)
    {
        if (file.Buffer.Length == 0)
            throw SijilliException.Validation("الملف فارغ أو غير صالح.", "Empty or invalid file.");
        if (file.Size > opts.MaxBytes)
            throw SijilliException.Validation(
                $"حجم الملف ({FormatBytes(file.Size)}) يتجاوز الحد المسموح ({FormatBytes(opts.MaxBytes)}).",
                $"File size {file.Size} exceeds limit {opts.MaxBytes}.");
        if (!opts.AllowedMimeTypes.Contains(file.MimeType))
            throw SijilliException.Validation(
                $"نوع الملف \"{file.MimeType}\" غير مدعوم.",
                $"Unsupported mime type {file.MimeType}.");

        var ext = SanitizeExtension(file.OriginalName, file.MimeType);
        var objectName = $"{Guid.NewGuid():N}{ext}";
        var relPath = $"{opts.PathPrefix.TrimEnd('/')}/{objectName}";
        var absPath = AbsoluteFor(opts.Bucket, relPath);

        Directory.CreateDirectory(Path.GetDirectoryName(absPath)!);
        // FileMode.CreateNew refuses to overwrite (parity with NestJS `flag: 'wx'`).
        await using (var fs = new FileStream(absPath, FileMode.CreateNew, FileAccess.Write))
            await fs.WriteAsync(file.Buffer, ct);

        return new UploadResult
        {
            Bucket = opts.Bucket,
            Path = relPath,
            Size = file.Size,
            MimeType = file.MimeType,
            Sha256 = Sha256Hex(file.Buffer),
        };
    }

    public async Task<byte[]> ReadAsync(string bucket, string path, CancellationToken ct)
    {
        var abs = AbsoluteFor(bucket, path);
        if (!File.Exists(abs)) throw SijilliException.NotFound("الملف", "File");
        return await File.ReadAllBytesAsync(abs, ct);
    }

    // Stream-friendly variant for large PDFs (avoids loading into memory).
    public Stream OpenRead(string bucket, string path)
    {
        var abs = AbsoluteFor(bucket, path);
        if (!File.Exists(abs)) throw SijilliException.NotFound("الملف", "File");
        return new FileStream(abs, FileMode.Open, FileAccess.Read, FileShare.Read);
    }

    public async Task<UploadResult> WriteRawAsync(string bucket, string path, byte[] data, string mime, CancellationToken ct)
    {
        var abs = AbsoluteFor(bucket, path);
        Directory.CreateDirectory(Path.GetDirectoryName(abs)!);
        await File.WriteAllBytesAsync(abs, data, ct);
        return new UploadResult
        {
            Bucket = bucket,
            Path = path,
            Size = data.LongLength,
            MimeType = mime,
            Sha256 = Sha256Hex(data),
        };
    }

    // Resolve and refuse path traversal: the candidate, after normalisation,
    // must remain inside <root>/<bucket>.
    private string AbsoluteFor(string bucket, string path)
    {
        var safeBucket = SafeBucketRe.Replace(bucket, "_");
        var bucketRoot = Path.GetFullPath(Path.Combine(_root, safeBucket));
        var candidate = Path.GetFullPath(Path.Combine(bucketRoot, path));
        var sep = Path.DirectorySeparatorChar;
        if (!candidate.StartsWith(bucketRoot + sep, StringComparison.OrdinalIgnoreCase) &&
            !candidate.Equals(bucketRoot, StringComparison.OrdinalIgnoreCase))
        {
            throw SijilliException.Validation("مسار غير صالح.", "Invalid path.");
        }
        return candidate;
    }

    private static string SanitizeExtension(string originalName, string mimeType)
    {
        var ext = Path.GetExtension(originalName).ToLowerInvariant();
        if (SafeExtRe.IsMatch(ext)) return ext;
        return mimeType switch
        {
            "application/pdf" => ".pdf",
            "image/jpeg" => ".jpg",
            "image/png" => ".png",
            _ => ".bin",
        };
    }

    private static string Sha256Hex(byte[] data) =>
        Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();

    private static string FormatBytes(long n) => n switch
    {
        < 1024 => $"{n} B",
        < 1024 * 1024 => $"{n / 1024.0:F1} KB",
        _ => $"{n / 1024.0 / 1024.0:F1} MB",
    };
}
